src/front/glsl/functions.rs

```diff
@@ -1351,4 +1336,143 @@ impl Program<'_> {
+                        base: coord,
+                        index: 0,
+                    },
+                    body,
+                ),
+                _ => coord,
+            };
+            let array_index = match arrayed {
+                true => {
+                    let index = match shadow {
+                        true => image_size.map_or(0, |s| s as u32 - 1),
+                        false => image_size.map_or(0, |s| s as u32),
+                    };
+
+                    Some(ctx.add_expression(Expression::AccessIndex { base: coord, index }, body))
+                }
+                _ => None,
+            };
+            let depth_ref = match shadow {
+                true => {
+                    let index = image_size.map_or(0, |s| s as u32);
+
+                    Some(ctx.add_expression(Expression::AccessIndex { base: coord, index }, body))
+                }
+                false => None,
+            };
+
+            Ok(CoordComponents {
+                coordinate,
+                depth_ref,
+                array_index,
+            })
+        } else {
+            Err(ErrorKind::SemanticError(
+                image_meta,
+                "Type is not an image".into(),
+            ))
+        }
+    }
+}
+
+/// Helper function to cast a expression holding a sampled a image to a
+/// depth one.
+///
+/// Creates a new expression to make sure the typifier doesn't return a
+/// cached evaluation.
+fn sampled_to_depth(
+    module: &mut Module,
+    ctx: &mut Context,
+    (image, meta): (Handle<Expression>, SourceMetadata),
+    body: &mut Block,
+) -> Result<Handle<Expression>, ErrorKind> {
+    let ty = match ctx[image] {
+        Expression::GlobalVariable(handle) => &mut module.global_variables.get_mut(handle).ty,
+        Expression::FunctionArgument(i) => {
+            ctx.depth_set.insert(i as usize);
+            &mut ctx.arguments[i as usize].ty
+        }
+        _ => {
+            return Err(ErrorKind::SemanticError(
+                meta,
+                "Not a valid texture expression".into(),
+            ))
+        }
+    };
+    match module.types[*ty].inner {
+        TypeInner::Image {
+            class,
+            dim,
+            arrayed,
+        } => match class {
+            ImageClass::Sampled { multi, .. } => {
+                *ty = module.types.fetch_or_append(Type {
+                    name: None,
+                    inner: TypeInner::Image {
+                        dim,
+                        arrayed,
+                        class: ImageClass::Depth { multi },
+                    },
+                })
+            }
+            ImageClass::Depth { .. } => {}
+            _ => return Err(ErrorKind::SemanticError(meta, "Not a texture".into())),
+        },
+        _ => return Err(ErrorKind::SemanticError(meta, "Not a texture".into())),
+    };
+
+    // Add a new expression to not have problems with the Typifier
+    // caching the old type
```

Comment:
This may be a bit of problem. So the new expressions are going to be typified correctly, but the old ones are still incorrect, and we are fine with this?

Reply:
It's definitely not the best solution and I wanted to discuss with you, how do you feel about adding a method to invalidate a handle in the typifier?

Reply:
yes, that sounds good